modules/webgl/src/utils/device-pixels.js

```diff
@@ -0,0 +1,42 @@
+/* global window */
+
+// multiplier need to convert CSS size to Device size
+export function cssToDeviceRatio(gl) {
+  return gl.drawingBufferWidth / (gl.canvas.clientWidth || gl.canvas.width);
+}
+
+// Maps CSS pixel position to device pixel position
+export function cssToDevicePixels(gl, cssPixel, yInvert = true) {
+  const ratio = cssToDeviceRatio(gl);
+  const width = gl.drawingBufferWidth;
+  const height = gl.drawingBufferHeight;
+  return scalePixels(cssPixel, ratio, width, height, yInvert);
+}
+
+// HELPER METHODS
+
+/**
+ * Calulates device pixel ratio, used during context creation
+ *
+ * @param {boolean or Number} useDevicePixels - boolean or a Number
+ * @return {Number} - device pixel ratio
+ */
+export function getDevicePixelRatio(useDevicePixels) {
+  const windowRatio = typeof window === 'undefined' ? 1 : window.devicePixelRatio || 1;
+  if (Number.isFinite(useDevicePixels)) {
+    return useDevicePixels <= 0 ? 1 : useDevicePixels;
+  }
+  return useDevicePixels ? windowRatio : 1;
+}
+
+// PRIVATE
+
+function scalePixels(pixel, ratio, width, height, yInvert) {
+  // since we are rounding to nearest, when dpr > 1, edge pixels may point to out of bounds value, clamp to the limit
+  const x = Math.min(Math.round(pixel[0] * ratio), width - 1);
+  const y = yInvert
+    ? Math.max(0, height - Math.ceil(ratio) - Math.round(pixel[1] * ratio))
```

Comment:
Why `Math.ceil(ratio)`? Is it not possible to get the bottom pixel if ratio = 2?

Reply:
`ceil` is only used for y invert cases, when pixel ratio is fraction, without `ceil` converted pixel location is not an integer. I have unit tests covering this case.

When pixel ratio is `2` , you can still select the bottom pixel, css 0 points to device 0, css 1 points to device 2 etc.